synapse/handlers/auth.py

```diff
@@ -1361,6 +1361,35 @@ async def start_sso_ui_auth(self, redirect_url: str, session_id: str) -> str:
             session = await self.store.get_ui_auth_session(session_id)
         except StoreError:
             raise SynapseError(400, "Unknown session ID: %s" % (session_id,))
+
+        user_id_to_verify = await self.get_session_data(
+            session_id, UIAuthSessionDataConstants.REGISTERED_USER_ID
+        )  # type: str
+
+        idps = await self.hs.get_sso_handler().get_identity_providers_for_user(
+            user_id_to_verify
+        )
+
+        if not idps:
+            # we checked that the user had some remote identities before offering an SSO
+            # flow, so either it's been deleted or the client has requested SSO despite
+            # it not being offered.
+            raise SynapseError(400, "User has no SSO identities")
```

Comment:
At this point in the flow should we be rendering JSON or HTML errors? (Should this use `sso_handler.render_error`?)

Reply:
good point.

Reply:
that said... there are plenty of other places we are raising SynapseErrors in this method. Ideally they'd all be caught and sent to `sso_handler.render_error` instead, and while we're at it I think `start_sso_ui_auth` probably just wants to move into `SsoHandler`, but that feels like a bigger refactor.